components/places/src/import/fennec_bookmarks.rs

```diff
@@ -0,0 +1,300 @@
+    );
+}
+
+const POPULATE_MIRROR_STRUCTURE: &str = "
+REPLACE INTO main.moz_bookmarks_synced_structure(guid, parentGuid, position)
+    SELECT stage.guid, stage.parent_guid, stage.pos FROM fennecBookmarksStaging stage;
+";
+
+lazy_static::lazy_static! {
+    static ref POPULATE_STAGING: &'static str =
+        "INSERT OR IGNORE INTO temp.fennecBookmarksStaging(
+            guid,
+            type,
+            parent_guid,
+            pos,
+            title,
+            bmkUri,
+            keyword,
+            tags,
+            date_added,
+            modified,
+            isLocal
+        )
+        SELECT
+            normalize_root_guid(b.guid),
+            b.type,
+            (SELECT normalize_root_guid(p.guid) FROM fennec.bookmarks p WHERE p._id = b.parent),
+            b.position,
+            b.title,
+            CASE
+                WHEN b.url IS NOT NULL
+                    THEN validate_url(b.url)
+                ELSE NULL
+            END,
+            b.keyword,
+            b.tags,
+            sanitize_timestamp(b.created),
+            sanitize_timestamp(b.modified),
+            1
+        FROM fennec.bookmarks b
+        WHERE NOT b.deleted
+        ;
+    ";
+
+    static ref CREATE_STAGING_TABLE: String = format!("
+        CREATE TEMP TABLE temp.fennecBookmarksStaging(
+            id INTEGER PRIMARY KEY,
+            guid TEXT NOT NULL UNIQUE,
+            type TINYINT NOT NULL
+                CHECK(type == {fennec_bookmark_type} OR type == {fennec_folder_type} OR type == {fennec_separator_type}),
+            parent_guid TEXT,
+            pos INT,
+            title TEXT,
+            bmkUri TEXT
+                CHECK(type != {fennec_bookmark_type} OR is_valid_url(bmkUri)),
+            keyword TEXT,
+            tags TEXT,
+            date_added INTEGER NOT NULL,
+            modified INTEGER NOT NULL,
+            isLocal TINYINT NOT NULL
+        )",
+        fennec_bookmark_type = FennecBookmarkType::Bookmark as u8,
+        fennec_folder_type = FennecBookmarkType::Folder as u8,
+        fennec_separator_type = FennecBookmarkType::Separator as u8,
+    );
+
+
+    static ref FIXUP_MOZ_BOOKMARKS: String = format!(
+        // Is there anything else?
+        "UPDATE main.moz_bookmarks SET
+           syncStatus = {unknown},
+           syncChangeCounter = 1,
+           lastModified = IFNULL((SELECT stage.modified FROM temp.fennecBookmarksStaging stage
+                                  WHERE stage.guid = main.moz_bookmarks.guid),
+                                 lastModified)",
+        unknown = SyncStatus::Unknown as u8
+    );
+}
+
+mod sql_fns {
+    use crate::import::common::sql_fns::{is_valid_url, sanitize_timestamp, validate_url};
+    use rusqlite::{functions::Context, Connection, Result};
+    use std::collections::HashMap;
+
+    lazy_static::lazy_static! {
+        static ref NORMALIZED_ROOT_GUIDS: HashMap<String, String> = {
```

Comment:
Nit: It doesn't matter much, but I think it's fine to use a slice of tuples here `&[("places", "root________"), ...]`, and iterate over them. Iterating over a static, 5-element slice is likely going to be faster than initializing a heap-allocated map and a hash lookup each time. Your call, though.